### DISCUSSION TITLE
Seeking guidance with project set up

### DISCUSSION THREAD

**Discussion by chrismiss1** [pointfreeco/swift-composable-architecture#1932] [Q&A]
Hi all,

I am fairly new to iOS and trying to get my hands dirty with TCA after watching most of the videos. What's not 100% clear to me is how I can start a new project and modularize the way they do in the videos to build modules independently. Are there any videos or blogs (that I couldn't find after searching) that walk you through how a project gets set up from scratch and how to add a feature/ module?

Thank you in advance!

**Comment by mbrandonw** [comment]
We have two episodes that cover the process from scratch:

[https://www.pointfree.co/episodes/ep171-modularization-part-1](https://www.pointfree.co/episodes/ep171-modularization-part-1)
[https://www.pointfree.co/episodes/ep172-modularization-part-2](https://www.pointfree.co/episodes/ep172-modularization-part-2)

Hopefully that helps.

**Reply by chrismiss1** [reply]
Yikes, @mbrandonw maybe I watched too many videos too quickly and forgot the earlier content... I don't think I can delete this post without admin privileges. Feel free to remove it :)

**Reply by mbrandonw** [reply]
No worries, someone else may have the same question so I think it's good to leave up. 🙂

**Reply by lightandshadow68** [reply]
This no longer seems to work in Xcode 15. 

When I create a new project, the directory structure looks like

Dashboard/Dashboard/Dashboard.xcodeproj

When I run 'swift package init'  in the root of the project (Dashboard) it creates the correct package files and folders on disk. And when I drag the folder root into the project in Xcode, I initially get a Dashboard package at the top level of the project. 

However, if I close the project, then reopen it, the reference seems to be lost, as Xcode displays it as a missing file in red. Deleting the broken reference and adding it again has the same result.

**Reply by mbrandonw** [reply]
Hi @lightandshadow68, it works for me if I first create the package, then create the app project inside the package, and then drag the root directory into the Xcode project.

**Reply by cdowney102** [reply]
I just recently created a new project and had the same issue. I retried a couple of times I think and not sure why or how but it worked on subsequent tries

**Reply by lightandshadow68** [reply]
Ahh.. that's the trick. The Xcode project should be inside the package. 

In the example, I thought the Swift package was initialized in the root of the Xcode project, such as inventory where...

> Inventory/SwiftUINavigation/SwiftUINavigation.xcodeproj

But, on closer inspection, It looks like the actual structure was...

> Inventory/SwiftUINavigation/SwiftUINavigation/SwiftUINavigation.xcodeproj

And the package was initialized one level up from the root of the Xcode package. 

Which works for me as well. 

Thanks!

**Reply by seanmrich** [reply]
The process is no longer working for me in Xcode 16. Dragging the root directory into Project Navigator will only allow a drop above the project which then prompts to create a workspace. Below is a step-by-step that used to work for me but breaks on the drag step. Anyone know a way to regain a modularized setup without going back to Xcode 15?

<details>
<summary>My previous process</summary>
* In Terminal, create the project folder Foo:

```
$ mkdir Foo
$ cd Foo
```

* Create the package:

```
$ swift package init
Creating library package: Foo
Creating Package.swift
...
```
* Create a directory for the project:
```
$ mkdir App
```
* In Xcode, create a new app project called Foo and put it in the App folder
* Open the Foo project
* Drag the top level Foo directory from Finder into the Xcode project and drop it immediately under the project name
* Close the Xcode project
* Create a file called Package.swift and place it in the App folder. Edit it to have an empty package content. This ensures Xcode won’t display that folder in the source navigator under the package header.
```swift
import PackageDescription

let package = Package(
  name: "",
  products: [],
  dependencies: [],
  targets: []
)
```
* Open the Xcode project. You should have top-level project name Foo. Under that will be the package, also named Foo. Then there will be the app sources, also named Foo. You can rename the app folder Foo to something like iOS if you want to have other targets for other platforms like macOS.

</details>

**Reply by bdhazman** [reply]
Having the same issue as @seanmrich. If anyone has any advice or workaround for this I'd love to hear!

**Reply by seanmrich** [reply]
I’ll add that updating to macOS 15 causes Xcode 15 to no longer launch. Still haven’t found a path forward. I looked for help on StackOverflow and Apple’s forums as well but no replies.

**Reply by bdhazman** [reply]
I found [this article](https://medium.com/@guycohendev/local-spm-mastering-modularization-with-swift-package-manager-xcode-15-e37b14c36199) that helped me. I don't think it's the exact setup as before but it seems to be about the same result.
